phpBB/phpbb/db/tools/doctrine.php

```diff
@@ -0,0 +1,867 @@
+			{
+				$this->_schema_create_index($column, $schema, $table_name, $index_name);
+			}
+		);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_index_drop(string $table_name, string $index_name)
+	{
+		return $this->_alter_schema(
+			function (Schema $schema) use ($table_name, $index_name): void
+			{
+				$this->_schema_index_drop($schema, $table_name, $index_name);
+			}
+		);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_create_unique_index(string $table_name, string $index_name, $column)
+	{
+		return $this->_alter_schema(
+			function (Schema $schema) use ($table_name, $index_name, $column): void
+			{
+				$this->_schema_create_unique_index($column, $schema, $table_name, $index_name);
+			}
+		);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_create_primary_key(string $table_name, $column)
+	{
+		return $this->_alter_schema(
+			function (Schema $schema) use ($table_name, $column): void
+			{
+				$this->_schema_create_primary_key($schema, $column, $table_name);
+			}
+		);
+	}
+
+	/**
+	 * Returns an array of indices for either unique and primary keys, or simple indices.
+	 *
+	 * @param string $table_name    The name of the table.
+	 * @param bool   $is_non_unique Whether to return simple indices or primary and unique ones.
+	 *
+	 * @return array The filtered index array.
+	 */
+	protected function get_filtered_index_list(string $table_name, bool $is_non_unique): array
+	{
+		try
+		{
+			$indices = $this->get_schema_manager()->listTableIndexes($table_name);
+		}
+		catch (Exception $e)
+		{
+			return [];
+		}
+
+		if ($is_non_unique)
+		{
+			return array_filter($indices, function (Index $index)
+			{
+				return $index->isSimpleIndex();
+			});
+		}
+
+		return array_filter($indices, function (Index $index)
+		{
+			return !$index->isSimpleIndex();
+		});
+	}
+
+	/**
+	 * Returns an array of lowercase asset names.
+	 *
+	 * @param array $assets Array of assets.
+	 *
+	 * @return array An array of lowercase asset names.
+	 */
+	protected function get_asset_names(array $assets): array
+	{
+		return array_map(
+			function (AbstractAsset $asset)
+			{
+				return strtolower($asset->getName());
+			},
+			$assets
+		);
+	}
+
+	/**
+	 * Returns whether an asset name exists in a list of assets (case insensitive).
+	 *
+	 * @param string $needle The asset name to search for.
+	 * @param array  $assets The array of assets.
+	 *
+	 * @return bool Whether the asset name exists in a list of assets.
+	 */
+	protected function asset_exists(string $needle, array $assets): bool
+	{
+		return in_array(strtolower($needle), $this->get_asset_names($assets), true);
+	}
+
+	/**
+	 * Alter the current database representation using a callback and execute the changes.
+	 * Returns false in case of error.
+	 *
+	 * @param callable $callback Callback taking the schema as parameters and returning it altered (or null in case of error)
+	 *
+	 * @return bool|string[]
+	 */
+	protected function _alter_schema(callable $callback)
```

Comment:
How attached are we to the underscore prefixed naming convention? Could we possibly kill it with fire?